### PR TITLE
[FIX] mail: check access on translation too

### PR DESCRIPTION
**Pull request by JKE-be**
This commit checks that users without the access right 'Template Editor'
are not allowed to add dynamic qweb in the translations of a mail
template.

Old tests have been updated to keep the assertRaises and all the code
that handle the invalidation cache, ... But now we ensure that the
Access Error raised is the expected one.

Original tests was testing that a user without the group 'Mail Editor'
got an  access error due to the missing group, but since the commit
https://github.com/odoo/odoo/commit/1a3e713c99474951cb5000c55a2382ba12074273 the Access Error is because you cannot edit a template that is
not created by yourself.

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
@std-odoo

File: addons/event/security/ir.model.access.csv [hunk 1]
```diff
@@ -26,3 +26,4 @@ access_event_tag,event.tag,model_event_tag,,0,0,0,0
 access_event_tag_registration,event.tag.user,model_event_tag,event.group_event_registration_desk,1,0,0,0
 access_event_tag_user,event.tag.user,model_event_tag,event.group_event_user,1,1,1,0
 access_event_tag_manager,event.tag.manager,model_event_tag,event.group_event_manager,1,1,1,1
+access_ir_model,access_ir_model,base.model_ir_model,event.group_event_user,1,0,0,0
```
Comment (odony) [inline addons/event/security/ir.model.access.csv:29]: might want to clarify in the PR desc that this commit is a separate bugfix ;)

File: addons/mail/models/ir_translation.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, _
+from odoo.exceptions import AccessError
+
+
+class IrTranslation(models.Model):
+    _inherit = 'ir.translation'
+
+    def _is_dynamic(self):
```
Comment (odony) [inline addons/mail/models/ir_translation.py:11]: it feels like we could avoid re-implementing this part, and delegate that check to the model itself, after browsing it with the appropriate context `lang`, in order to check the translated value

File: addons/mail/models/ir_translation.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        return False
+
+    def check(self, mode):
+        """ Check access rights of operation ``mode`` on ``self`` for the
+        current user. Raise an AccessError in case conditions are not met.
+        """
+        super().check(mode)
+
+        # if we don't modify translation of mail.template, we ignore
+        translations_for_mail_template = self.filtered(lambda template: template.name.startswith('mail.template,'))
```
Comment (odony) [inline addons/mail/models/ir_translation.py:31]: we probably need to handle all `mail.render.mixin` models here, and if we delegate the check to the actual model, it will take care of honoring the `_unrestricted_rendering` flag etc.

perf-wise it should be ok, we don't edit those translations all the time

File: addons/mail/tests/test_mail_template.py [hunk 4]
```diff
@@ -46,6 +46,9 @@ def test_mail_compose_message_content_from_template_mass_mode(self):
         self.assertIn('13', values[self.partner_employee.id]['body_html'], 'We must trust mail template values')
 
     def test_mail_template_acl(self):
+        EXC_SELF_MSG = "Due to security restrictions, you are not allowed to modify 'Email Templates' (mail.template) records."
+        EXC_DYN_MSG = 'Only users belonging to the "Mail Template Editor" group can modify dynamic templates.'
```
Comment (odony) [inline addons/mail/tests/test_mail_template.py:50]: as discussed live, we could avoid the specific matching of exception by doing the steps in the right order.

The problem with the previous tests was that we could confuse an AccessError caused by the ownership of the template with an error caused by dynamic code being present. We can avoid that doubt by always having each "dynamic code edit" step preceded by a "static code edit" step *on the same template*, with *the same user*. With good comments to explain that this is important ;-)

That being said, I don't mind the extra specificity of the exception messages, it just makes the tests a bit more verbose/fragile.

**Review by odony — COMMENTED** [review]
first quick look :)

**Review by std-odoo — COMMENTED** [review]
Rivou :) 

I wonder if it's really robust, I know that for some short words (like "yes", "no",...) it can take the translation defined for something else, not 100% how it works but it might be worth to check

File: addons/event/security/ir.model.access.csv (hunk 1 above)
Comment (tde-banana-odoo) [inline addons/event/security/ir.model.access.csv:29]: I don't understand this fix

'A event user need to be able to create/edit  mail template for the
event.registration.
So the read access on ir.model is needed to be able to edit the mail
template.'

What do you mean by that ? Either there is an issue with model when people create templates (which is something doable outside of event) and we have to try to sudo access to ir.model, either you wanted to highlight a specific flow ?

Comment (tde-banana-odoo) [inline addons/event/security/ir.model.access.csv:29]: I don't understand this fix

'A event user need to be able to create/edit mail template for the
event.registration.
So the read access on ir.model is needed to be able to edit the mail
template.'

What do you mean by that ? Either there is an issue with model when people create templates (which is something doable outside of event) and we have to try to sudo access to ir.model, either you wanted to highlight a specific flow ?

File: addons/mail/models/ir_translation.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        self._check_is_dynamic()
+        return res
+
+    def _check_is_dynamic(self):
+        # if we don't modify translation of at least a model that inherits from mail.render.mixin, we ignore it
+        translations_for_mail_render_mixin = self.filtered(lambda tmpl: isinstance(self.env[tmpl.name.split(',')[0]], self.pool['mail.render.mixin']))
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:24]: tmpl is misleading, especially in this fix, I was wondering why you were browsing templates on translation model. Maybe rename tmpl to translation ?

Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:24]: instead of isinstance shouldn't we use subclass ?

File: addons/mail/models/ir_translation.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            return
+
+        # if we are admin, or that we can update mail.template we ignore
+        if self.env.is_admin() or self.env.user.has_group('mail.group_mail_template_editor'):
+            return
+
+        # Check that we don't add qweb code in translation when you don't have the rights
+        for trans in translations_for_mail_render_mixin:
+            res_model, res_id = trans.name.split(',')[0], trans.res_id
+            rec = self.env[res_model].with_context(lang=trans.lang).browse(res_id)
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:35]: Could you batch by model / res_ids ? Otherwise you are doing a single-record based computation in a method that is intended to be done in batch. This implies a bit of preparation but in the end it is better to batch when possible.

File: addons/mail/models/ir_translation.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+            if rec._is_dynamic():
+                group = self.env.ref('mail.group_mail_template_editor')
+                raise AccessError(
+                    _('Only users belonging to the "%s" group can modify translation related to dynamic templates.', group.name)
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:40]: Could be great to give more details in error message (which template failed ?). Otherwise sometimes you end up with something you can't fixbecause you don't really know which records are failing.

File: addons/mail/tests/test_mail_template.py [hunk 8]
```diff
@@ -70,12 +70,53 @@ def test_mail_template_acl(self):
...
         with self.assertRaises(AccessError):
             mail_template.with_user(self.user_employee).body_html = '<p t-esc="\'foo\'"></p>'
 
+        # Standard employee can edit his own templates if not dynamic
+        employee_template.with_user(self.user_employee).body_html = '<p>foo</p>'
+
         # Standard employee cannot create and edit templates with dynamic inline fields
         with self.assertRaises(AccessError):
-            self.env['mail.template'].with_user(self.user_employee).create({'email_to': '{{ object.partner_id.email }}'})
+            employee_template.with_user(self.user_employee).create({'email_to': '{{ object.partner_id.email }}'})
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:84]: Why this change ? Seems not necessary, as anyway we are calling create ?

File: addons/mail/tests/test_mail_template.py [hunk 9]
```diff
@@ -70,12 +70,53 @@ def test_mail_template_acl(self):
...
+
+        # Standard employee cannot edit his own templates if dynamic
+        with self.assertRaises(AccessError):
+            employee_template.with_user(self.user_employee).body_html = '<p t-esc="\'foo\'"></p>'
+
+        with self.assertRaises(AccessError):
+            employee_template.with_user(self.user_employee).email_to = '{{ object.partner_id.email }}'
+
+    def test_mail_template_acl_translation(self):
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:93]: Could you add a docstring explaining the purpose and expected results of the test ?

File: addons/mail/models/ir_translation.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, _
+from odoo.exceptions import AccessError
+
+
+class IrTranslation(models.Model):
+    _inherit = 'ir.translation'
+
+
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:11]: Linting: too much void lines

File: addons/mail/models/ir_translation.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    def create(self, vals_list):
+        res = super().create(vals_list)
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:13]: could rename res to translations, indicating this is a recordset

**Review by tde-banana-odoo — COMMENTED** [review]
Validation (code mainly, still did not test)

File: addons/event/security/ir.model.access.csv (hunk 1 above)
Comment (JKE-be) [reply addons/event/security/ir.model.access.csv:29]: This fix is when you start a clean db with just event installed.
If you try to create a event, with mail template for event registration it will crash because you need to have read access on ir.model to be able to set the mail template for model 'event.registration' (the default value)

File: addons/mail/models/ir_translation.py (hunk 5 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py:24]: :+1: , previously I only check that is was translation of mail.template and not (sub)class of mail.render.mixin, so it was a leftover wrong name.

Renamed to translation

File: addons/event/security/ir.model.access.csv (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/event/security/ir.model.access.csv:29]: The issue comes probably from template_ref reference field, that is using the model field in its options. I think we should fix the reference field, as I don't think we have to give access to ir.model just for the implementation of reference fields. Pinging @mart-e for this one, as he removed access on ir.model, not sure how we intend to use reference fields (using model a bit with magic).

File: addons/mail/models/ir_translation.py (hunk 5 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py:24]: both works, but yes, I can change it, more readable :+1: 
and more coherent with the code of `_reflect_model_params`

File: addons/event/security/ir.model.access.csv (hunk 1 above)
Comment (mart-e) [reply addons/event/security/ir.model.access.csv:29]: Indeed it's better if we can fix the (web) framework to avoid this call instead of adding the access, it should not be required.

File: addons/mail/models/ir_translation.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            for res_model, res_ids in ids_by_model_by_lang[lang].items():
+                self.env[res_model].with_context(lang=lang).browse(res_ids)
+
+        for trans in translations_for_mail_render_mixin:
+            res_model, res_id = trans.name.split(',')[0], trans.res_id
+            rec = self.env[res_model].with_context(lang=trans.lang).browse(res_id)
+
+            if rec._is_dynamic():
+                group = self.env.ref('mail.group_mail_template_editor')
+                more_info = len(self) > 1 and ' [%s]' % rec or ''
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:53]: Not sure what this more_info is about, are we trying to display the related record ? Shouldn't we add either the ID, either the display_name ?

File: addons/mail/models/ir_translation.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                raise AccessError(
+                    _('Only users belonging to the "%(group)s" group can modify translation related to dynamic templates.%(xtra)s',
+                        group=group.name,
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_translation.py:56]: Linting: should be aligned vertically with the ``'`` I think as they are all parameters of the call to ``_`` (such review)

File: addons/mail/tests/test_mail_template.py [hunk 14]
```diff
@@ -70,12 +70,57 @@ def test_mail_template_acl(self):
...
+        ''' Test that a user that doenn't have the group_mail_template_editor cannot create / edit
+        translation with dynamic code if he cannot write dynamic code on the related record itself.
+        '''
+
+        self.env.ref('base.lang_fr').sudo().active = True
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:98]: Shouldn't we use ``_activate_lang`` (and ``_update_translations`` maybe) ?

File: addons/mail/tests/test_mail_template.py [hunk 15]
```diff
@@ -70,12 +70,57 @@ def test_mail_template_acl(self):
...
+
+        employee_template = self.env['mail.template'].with_user(self.user_employee).create({'body_html': '<p>foo</p>'})
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:100]: Could you set a model (to have a coherent template) ? And maybe a subject so that we also that we try inline translated fields (in addition to qweb).

File: addons/mail/tests/test_mail_template.py [hunk 16]
```diff
@@ -70,12 +70,57 @@ def test_mail_template_acl(self):
...
+
+        Translation = self.env['ir.translation']
+        Translation.insert_missing(employee_template._fields['body_html'], employee_template)
+        employee_translations = Translation.with_user(self.user_employee).search(
+            [('res_id', '=', employee_template.id), ('name', '=', 'mail.template,body_html'), ('lang', '=', 'fr_FR')]
+        )
+        # keep a copy to create new translation later
+        create_vals = employee_translations[0].read([])[0]
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:108]: body_translation_vals ?

File: addons/mail/tests/test_mail_template.py [hunk 17]
```diff
@@ -70,12 +70,57 @@ def test_mail_template_acl(self):
...
+
+        # write on translation for template without dynamic code is allowed
+        employee_translations[0].value = 'non-qweb'
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:111]: we could probably remove the [0] (maybe add a limit=1 in the search but that is probably not necessary)

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for updating :) some additional comments, mainly on small remarks or linting on tests.

File: addons/mail/models/ir_translation.py (hunk 12 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py:53]: The display name is not interesting in this case, and the id not enough, what you want to know is which record is blocking your request.
It is what we do here with rec. It will show the browser record, and some the model and the id that is blocking, since is can be another model that mail.template.

File: addons/mail/tests/test_mail_template.py (hunk 15 above)
Comment (JKE-be) [reply addons/mail/tests/test_mail_template.py:100]: I can add model and co, but same for all previous tests in this case :thinking: ... 

About the inline, I can add a test again if you really want, but it is already tested by previous test `test_mail_template_acl`.
This one, just test that the _is_dynamic call is done on translation create/write too.

File: addons/mail/models/ir_translation.py (hunk 13 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py:56]: old indent was not wrong but if you prefer... even if perso I really prefer a indent of 4.
it's porbably because the first param (the string) is not at the line that it shock your eyes ;) but it will become ugly
image

Comment (tde-banana-odoo) [reply addons/mail/models/ir_translation.py]: My linter wants me to align everything ... but I think it is a bit outdated, as some valid constructs are considered invalid ^^ . Anyway this is not the most important part of the review, as you prefer :D .

File: addons/mail/tests/test_mail_template.py (hunk 15 above)
Comment (tde-banana-odoo) [reply addons/mail/tests/test_mail_template.py:100]: Well indeed, maybe there are other instances but always better to at least give a model. The field is not required, but it may become required someday (as it makes no sense to have generic templates), better be already compliant :) .

For subject, indeed it is somehow already tested (at mixin / template level), but as you test the manual write on the translation for body, we could also test the manual write on translation for subject. That way we test the mixin, and the translations, at both levels.

Comment (JKE-be) [reply addons/mail/tests/test_mail_template.py]: done

File: addons/mail/models/ir_translation.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        # if we don't modify translation of at least a model that inherits from mail.render.mixin, we ignore it
+        # translation.name can be a path, and so not in the pool, so type(None) will exclude these translations.
+        translations_for_mail_render_mixin = self.filtered(
+            lambda translation: issubclass(type(self.env.get(translation.name.split(',')[0])), self.pool['mail.render.mixin'])
+        )
+        if not translations_for_mail_render_mixin:
+            return
+
+        # if we are admin, or that we can update mail.template we ignore
+        if self.env.is_admin() or self.env.user.has_group('mail.group_mail_template_editor'):
```
Comment (std-odoo) [inline addons/mail/models/ir_translation.py:31]: can we move this at the beginning of the method ?
(so we skip everything if we are admin)

**Review by std-odoo — COMMENTED** [review]
@JKE-be One last thing, otherwise LGTM :)

File: addons/mail/models/ir_translation.py (hunk 18 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py]: I did it, but now we will do query has_group (probably already in cache) on each create/write before to check if we write on field that require this group :thinking: 

I don't think that the issubclass is heavy... 
Not sure that the move is smart. what do you think about ?

Comment (std-odoo) [reply addons/mail/models/ir_translation.py]: hum indeed :thinking: but in the filtered you do translation.name, but it should be in the cache as well
both are ok for me

**Review by tde-banana-odoo — APPROVED** [review]
Feature and tests LGTM, thanks @JKE-be for the followup :)

File: addons/mail/models/ir_translation.py (hunk 18 above)
Comment (JKE-be) [reply addons/mail/models/ir_translation.py]: translation is already in cache since it is called after the create and the write ;)
So yes, I will move it back after, since it is ok for you too.
It will avoid a random extra query on each write of ir.translation.

**Comment by tde-banana-odoo** [comment]
@robodoo r+
